man/ipetoipe.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.TH IPETOIPE 1 "June 27, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
ipetoipe \- Convert between Ipe file formats
.SH SYNOPSIS
.B ipetoipe
( -pdf | -eps | -xml ) { options } \fIinput-file\fP [ \fIoutput-file\fP ]

.SH DESCRIPTION
.PP
\fBipetoipe\fP converts between the three Ipe file formats XML, PDF,
and EPS.   Options are:
.TP
\fB-pdf\fP
convert to PDF format
.TP
\fB-eps\fP
convert to EPS (encapsulated Postscript) format
.TP
\fB-xml\fP
convert to XML format
.TP
\fB-export\fP
do not include Ipe markup in the output file.
.br
\fIWarning:\fP this means that Ipe will not be able to read the
resulting output file. 
.TP
\fB-pages\fP \fIfrom\fP-\fIto\fP
restrict exporting to PDF to this page range.  This implies  the
\fB-export\fP option. 
.TP
\fB-view\fP \fIpage\fP-\fIview\fP
only export this single view from the document.  This implies  the
\fB-export\fP option. 
.TP
\fB-markedview\fP
create a Postscript/PDF page only for the marked views of marked Ipe
pages. This is for instance convenient for making handouts for slides used in
a presentation. This implies the
\fB-export\fP option. 
.TP
\fB-runlatex\fP
run Latex even for XML output. This has the effect of including the
dimensions of each text object in the XML file.
.TP
\fB-nocolor\fP
do not use any color commands in Postscript output.  This is useful if
you are preparing a manuscript for publication where the publisher
checks that the figures are grayscale only.  Note that this option
doesn't actually verify that the figure contains no color (a figure
using color processed with this option will cause a Postscript error
when displayed).
.TP
\fB-nozip\fP
do not compress streams in PDF or Postscript output.

.SH ENVIRONMENT VARIABLES

\fBipetoipe\fP respects the following environment variables:

.TP
\fBIPELATEXDIR\fP
the directory where \fBipetoipe\fP runs Pdflatex.

.TP
\fBIPEDEBUG\fP
set this to 1 for debugging output.

.SH AUTHOR
Otfried Cheong

.SH REPORTING BUGS
.ad l
Please report bugs using Ipe bugzilla at
.I "http://ipe7.sourceforge.net/bugzilla.html"

.SH SEE ALSO
.ad l
Ipe is documented fully in
.IR "The Ipe Manual" ,
which is available online at
.I "http://ipe7.sourceforge.net/manual/manual.html" .


news.txt
--------------------------------------------------------------------
 
Ipe 7.1.1 (2011-09-07) SVN 563 Oberwolfach edition

 * Fixed crash on odoc event on Mac OS X (bug #63).

 * When a point is created using the lines tool (by clicking left and
   then right at the same location), force line cap to be round to
   avoid making invisible object (and for consistency with ink tool).

 * IPESTYLES is now a list of style directories.  Ipe will search for
   stylesheets in all these directories. By default, this will now
   include ~/.ipe/styles on Unix, and also ~/Library/Ipe/Styles on
   OSX.

 * Fixed omission of ~/.ipe/ipelets on OSX and added
   ~/Library/Ipe/Ipelets as well.

 * Layers can now be moved in the layerlist (right-click on layer
   name).

 * Zoom and pan now properly initialized when Ipe starts up (feature
   #4).

 * New script "add-style" to add stylesheet to an Ipe figure.

 * Removed "-styleupdate" option from ipetoipe, and created new script
   "update-styles".  Updating styles both inside Ipe and using
   "update-styles" now searches the entire IPESTYLES list.

 * Reenabled compilation of omitted image ipelet (bug #66).

 * New <textpad> setting in stylesheet to fix cropped text (bug #49).

 * Always present 'normal' as the first attribute in dropdown list.

 * Show real linewidth during handwriting.

--------------------------------------------------------------------
 
Ipe 7.1.0 (2011-08-29) SVN 549

 * Refactored Ipe code so that it is easy to build the Ipe application
   with different user interface toolkits.  A native Win32 interface
   is nearly complete.  A Gtk interface is being considered, with
   Canvas and dialog support implemented.  Perhaps someone will
   contribute a native Cocoa interface for Mac OS X.

 * Selection preserved after undo.

 * When adding objects to the selection, the last selected object is
   the primary selection.

 * New shredder mode (feature #16).

 * Easier customization of mouse buttons using a table in "mouse.lua".
   More than three buttons can now be used.  New "shred" operation.

 * Added "ipescript" tool.  It is no longer necessary to install the
   Ipe module for Lua to run scripts.

 * Fixed bounding box when exporting a view to Postscript.

 * Properties menu now also shown when no object selected, and allows
   to change properties of multiple objects at once.

 * Updated Makefile and instructions for compiling on MacOS X (bug
   #20). Including shell script for building an app bundle by leahcimb.

 * Fixed regular k-gon ipelet (bug #35).

 * Fixed prefs.lua for MacOS X (bug #28).

 * Layers are now again presented in the order in which they were
   added (bug #32).  In the future we'll need a feature to rearrange
   them manually.

 * Added preference for zoom factor (feature #53).

 * On MacOS X, the close button should now correctly reflect whether
   the document is modified.

 * Fixed display of subsection bookmarks (bug #31).

 * Fixed spurious loss of pinning information for group objects (bug #37).

 * First attempt to handle file-open events on Mac OS X.
   This would allow to launch documents by clicking them (feature #58).

 * Shortcut to rename the active layer (feature #50).

 * If prefs.tablet_mode is true, then in ink mode the selection is not
   highlighted and the cursor remains a dot (feature #26).
 
--------------------------------------------------------------------
 
Ipe 7.0.14 (2011-01-27) SVN 474 Australia edition

 * Fixed saving of <pagenumberstyle> (bug #7).

 * Fixed iperender and thumbnails on 64-bit platforms.

 * Pressing Enter in view/page selector now jumps to the view/page
   (feature #24).  Shortcuts for view/page selector.  Original
   view/page highlighted in selector.

 * Background no longer hides page number in PDF output (bug #8).

 * Mark circle center now also works for arcs (bug #3).

 * Fit to selection/objects now fits to visible object, not including
   control points (bug #2).

 * Grid now only drawn inside page's frame area, drawn grid can be
   coarser than grid resolution, thick lines drawn at configurable
   steps (see prefs.lua)  (feature #17).

 * Correctly render Latex text inside symbols (bug #18).

 * matrix:inverse() no longer crashes on singular matrix.

 * Circle to ellipse transformation no longer crashes on singular
   ellipse (bug #21).

 * Easier to use symbols by remembering current symbol, which can then
   be inserted with one function key (Alt+Y) (feature #19).

 * Introduced pref.close_distance.

 * Paste at cursor now correctly observes pinning of the object.

 * Fixed serious memory leak in group object.

 * Implemented marking of pages and views for printing in Ipe, and
   selecting those in ipetoipe.

--------------------------------------------------------------------

Ipe 7.0.13 (2010-11-26) SVN 466

 * Error message when document requires charset conversion but iconv
   is not compiled in.

 * Fixed selection of invisible objects.

 * New dialog for jumping to a page or view, showing thumbnails of the
   pages/ views.
   This is implemented using a new widget and function in
   libipeqtcanvas so that it can be reused easily:
   int pno = ipeqt::PageSelector::selectPageOrView(doc)

 * When using the ink tool, actual stroke color is used while
   handwriting. 

 * Pressing the eraser end of a pen should generate a middle button
   event.

 * New action: "fit to page width" 

 * In handwriting mode, cursor is a dot.

 * New preference: prefs.coordinates_format

 * Feedback in status line for line and spline tools now permanent.

 * Pressing the pen without movement in ink tool now makes a dot.

 * Handwriting now creates paths with round line cap, regardless of
   current setting (but that can still be changed afterwards).

 * Safety check for reasonable locale on startup.

 * Layers now shown in the order in which they appear in views of the
   page. 

 * Page sorter that lets you rearrange the pages of the document.


src/ipetoipe/ipetoipe.cpp
// --------------------------------------------------------------------
// ipetoipe
// --------------------------------------------------------------------
/*

    This file is part of the extensible drawing editor Ipe.
    Copyright (C) 1993-2011  Otfried Cheong

    Ipe is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    As a special exception, you have permission to link Ipe with the
    CGAL library and distribute executables, as long as you follow the
    requirements of the Gnu General Public License in regard to all of
    the software in the executable aside from CGAL.

    Ipe is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
    License for more details.

    You should have received a copy of the GNU General Public License
    along with Ipe; if not, you can find it at
    "http://www.gnu.org/copyleft/gpl.html", or write to the Free
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

#include "ipedoc.h"

using ipe::Document;
using ipe::String;

static int topdforps(Document *doc, String src, String dst,
		     Document::TFormat fm, uint flags,
		     int fromPage = -1, int toPage = -1, int viewNo = -1)
{
  if (fm == Document::EEps) {
    if (viewNo < 0 && doc->countTotalViews() > 1) {
      fprintf(stderr, "The document contains %d views, "
	      "cannot convert to EPS format.\n", doc->countTotalViews());
      return 1;
    }

    bool trans = doc->hasTransparency();
    bool grad = doc->hasGradients();

    if (!trans && !grad) {
      int res = doc->runLatex();
      if (res)
	return res;
    }

    bool ttf = doc->hasTrueTypeFonts();

    if (trans || grad || ttf) {
      fprintf(stderr, "The document uses the following features:\n");
      if (trans)
	fprintf(stderr, " - transparency\n");
      if (grad)
	fprintf(stderr, " - gradients\n");
      if (ttf)
	fprintf(stderr, " - Truetype fonts\n");
      fprintf(stderr,
	      "Ipe does not support these features for Postscript output.\n"
	      "You can save as PDF, and use pdftops to create a "
	      "Postscript file.\n");
      return 1;
    }
  } else {
    int res = doc->runLatex();
    if (res) return res;
  }

  bool result = false;
  if (viewNo >= 0) {
    result = doc->exportView(dst.z(), fm, flags, fromPage, viewNo);
  } else if (toPage >= 0) {
    result = doc->exportPages(dst.z(), flags, fromPage, toPage);
  } else {
    result = doc->save(dst.z(), fm, flags);
  }
  if (!result) {
    fprintf(stderr, "Failed to save or export document!\n");
    return 1;
  }

  if (flags & Document::EExport)
    fprintf(stderr, "Warning: the exported file contains no Ipe markup.\n"
	    "It cannot be read by Ipe - make sure you keep the original!\n");
  return 0;
}

static void usage()
{
  fprintf(stderr,
	  "Usage: ipetoipe ( -xml | -eps | -pdf ) <options> "
	  "infile [ outfile ]\n"
	  "Ipetoipe converts between the different Ipe file formats.\n"
	  " -export      : output contains no Ipe markup.\n"
	  " -pages <n-m> : export only these pages (implies -export).\n"
	  " -view <p-v>  : export only this view (implies -export).\n"
	  " -markedview  "
	  ": export only marked views on marked pages (implies -export).\n"
	  " -runlatex    : run Latex even for XML output.\n"
	  " -nocolor     : avoid any color commands in EPS output.\n"
	  " -nozip:      : do not compress PDF streams.\n"
	  );
  exit(1);
}

int main(int argc, char *argv[])
{
  ipe::Platform::initLib(ipe::IPELIB_VERSION);

  // ensure at least two arguments (handles -help as well :-)
  if (argc < 3)
    usage();

  Document::TFormat frm = Document::EUnknown;
  if (!strcmp(argv[1], "-xml"))
    frm = Document::EXml;
  else if (!strcmp(argv[1], "-eps"))
    frm = Document::EEps;
  else if (!strcmp(argv[1], "-pdf"))
    frm = Document::EPdf;

  if (frm == Document::EUnknown)
    usage();

  uint flags = Document::ESaveNormal;
  bool runLatex = false;
  int fromPage = -1;
  int toPage = -1;
  int viewNo = -1;
  int i = 2;

  String infile;
  String outfile;

  while (i < argc) {

    if (!strcmp(argv[i], "-export")) {
      flags |= Document::EExport;
      ++i;
    } else if (!strcmp(argv[i], "-view")) {
      if (sscanf(argv[i+1], "%d-%d", &fromPage, &viewNo) != 2)
	usage();
      flags |= Document::EExport;
      i += 2;
    } else if (!strcmp(argv[i], "-pages")) {
      if (sscanf(argv[i+1], "%d-%d", &fromPage, &toPage) != 2)
	usage();
      flags |= Document::EExport;
      i += 2;
    } else if (!strcmp(argv[i], "-markedview")) {
      flags |= Document::EMarkedView;
      flags |= Document::EExport;
      ++i;
    } else if (!strcmp(argv[i], "-runlatex")) {
      runLatex = true;
      ++i;
    } else if (!strcmp(argv[i], "-nocolor")) {
      flags |= Document::ENoColor;
      ++i;
    } else if (!strcmp(argv[i], "-nozip")) {
      flags |= Document::ENoZip;
      ++i;
    } else {
      // last one or two arguments must be filenames
      infile = argv[i];
      ++i;
      if (i < argc) {
	outfile = argv[i];
	++i;
      }
      if (i != argc)
	usage();
    }
  }

  if (infile.empty())
    usage();

  if ((flags & Document::EExport) && frm == Document::EXml) {
    fprintf(stderr, "-export only available with -pdf and -eps.\n");
    exit(1);
  }

  if (toPage >= 0 && frm != Document::EPdf) {
    fprintf(stderr, "-pages only available with -pdf.\n");
    exit(1);
  }

  if (toPage >= 0 && viewNo >= 0) {
    fprintf(stderr, "cannot specify both -pages and -view.\n");
    exit(1);
  }

  if (outfile.empty()) {
    outfile = infile;
    String ext = infile.right(4);
    if (ext == ".ipe" || ext == ".pdf" || ext == ".xml" || ext == ".eps")
      outfile = infile.left(infile.size() - 4);
    switch (frm) {
    case Document::EXml:
      outfile += ".ipe";
      break;
    case Document::EPdf:
      outfile += ".pdf";
      break;
    case Document::EEps:
      outfile += ".eps";
    default:
      break;
    }
    if (outfile == infile) {
      fprintf(stderr, "Cannot guess output filename.\n");
      exit(1);
    }
  }

  IpeAutoPtr<Document> doc(Document::loadWithErrorReport(infile.z()));

  if (!doc.ptr())
    return 1;

  fprintf(stderr, "Document %s has %d pages (%d views)\n",
	  infile.z(), doc->countPages(), doc->countTotalViews());

  // check frompage, topage, viewno
  if (toPage >= 0) {
    if (fromPage <= 0 || fromPage > toPage || toPage > doc->countPages()) {
      fprintf(stderr, "incorrect -pages specification.\n");
      exit(1);
    }
    --fromPage;
    --toPage;
  } else if (fromPage >= 0) {
    if (fromPage <= 0 || fromPage > doc->countPages() ||
	viewNo <= 0 || viewNo > doc->page(fromPage - 1)->countViews()) {
      fprintf(stderr, "incorrect -view specification.\n");
      exit(1);
    }
    --fromPage;
    --viewNo;
  }

  char buf[64];
  sprintf(buf, "ipetoipe %d.%d.%d",
	  ipe::IPELIB_VERSION / 10000,
	  (ipe::IPELIB_VERSION / 100) % 100,
	  ipe::IPELIB_VERSION % 100);
  Document::SProperties props = doc->properties();
  props.iCreator = buf;
  doc->setProperties(props);

  switch (frm) {
  case Document::EXml:
    if (runLatex)
      return topdforps(doc.ptr(), infile, outfile, Document::EXml, flags);
    else
      doc->save(outfile.z(), Document::EXml, Document::ESaveNormal);
  default:
    return 0;

  case Document::EPdf:
    return topdforps(doc.ptr(), infile, outfile, Document::EPdf,
		     flags, fromPage, toPage, viewNo);

  case Document::EEps:
    return topdforps(doc.ptr(), infile, outfile, Document::EEps,
		     flags, fromPage, toPage, viewNo);
  }
}

// --------------------------------------------------------------------
